### PR TITLE
Support hparams logging to tensorboard 

### LINKED ISSUE #428: Also log hyperparameters to the tensorboard
### 🚀 Feature

It would be nice if the used hyperparameters would be included in the tensorboard log for better comparability.

### Motivation

Manually comparing the hyperparameters in a separate location isn't as clean and tensorboard supports it. This would use the full potential.

### Additional context

https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter.add_hparams

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by timothe-chaumont**
## Description

- [x] I created a `HParam` data class in the same way as `Figure`, `Image` ones. It can take any number of distinct hyperparameters and metrics as input.

- [x] In each `.write()` method of `HumanOutputFormat`, `JSONOutputFormat` and `CSVOutputFormat`, I have raised errors when `hparams` format was given. _e.g._
```python
            elif isinstance(value, HParam):
                raise FormatUnsupportedError(["stdout", "log"], "hparam")
```

- [x] I added a case in `TensorBoardOutputFormat`  to log the hparams values to tensorboard:
```python
            if isinstance(value, HParam):
                # we don't use `self.writer.add_hparams` to have control over the log_dir
                exp, ssi, sei = hparams(value.hparam_dict, metric_dict=value.metric_dict)
                self.writer.file_writer.add_summary(exp)
                self.writer.file_writer.add_summary(ssi)
                self.writer.file_writer.add_summary(sei)
```
As  described in the code comment, I have not used `self.writer.add_hparams(hparam_dict, metric_dict)`, provided by [pytorch](https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter.add_hparams), but reused [some of the code](https://pytorch.org/docs/stable/_modules/torch/utils/tensorboard/writer.html#SummaryWriter.add_hparams) from that method. Using this first method, the hparams could not be saved in the same _run_ folder as the other logs, and so metrics from SCALARS tab could not appear in HPARAMS tab.

- [x] I created one parametrized test : `test_report_hparam_to_unsupported_format_raises_error()` in the same way as other data classes. I did not create a `test_report_hparam_to_tensorboard` test, as hparams logs are not seen by `EventAccumulator` (c.f. [Stack Overflow](https://stackoverflow.com/questions/70033253/read-out-tensorboard-hparams-programatically) and [EventAccumulator implementation](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/event_accumulator.py)).

- [x] Finally, in the tensorboard section of the documentation, I added an example of a callback that uses this new code to log hyperparameters to tensorboard. 

## Additionnal information - choices made
- As hyperparameters are key-value pairs, adding the support for csv, json, or human output formats could be a future developpment.
- It is not required to pass a `metric_dict` to `hparams()` or `writer.add_hparams()`, but if we don't, then nothing is displayed in HPARAMS tab. I have added a warning to alert the user about that.
- When adding metrics in `metric_dict`, the users have 2 choices:
  - Adding custom metrics -> only one value per run for each metric and no plot,
  -  Using metrics from the scalar section (with value 0) ->  the user can visualize the plot of that metric in HPARAMS tab and the last value of that plot is displayed (c.f. image below).

  It is not ideal to display the last value - instead of the best value, for example (issue discussed [here](https://github.com/tensorflow/tensorboard/issues/3755)) - but I decided to use the second one in the documentation example as I found it more relevant and intuitive.

- In the example, I put the logic in `_on_training_start()` as we only have to log the hyperparameters & metrics once, but I could also added it in `on_step()` as I did it [here](https://github.com/hill-a/stable-baselines/issues/1128#issuecomment-1124794750).

tensorboard HPARAMS tab screenshot

## Motivation and Context

- ~~[x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)~~ I have proposed a solution to #428 : a way of logging hyperparameters to tensorboard.

closes #428 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/logger.py [hunk 1]
```diff
@@ -389,6 +414,13 @@ def write(self, key_values: Dict[str, Any], key_excluded: Dict[str, Union[str, T
             if isinstance(value, Image):
                 self.writer.add_image(key, value.image, step, dataformats=value.dataformats)
 
+            if isinstance(value, HParam):
+                # we don't use `self.writer.add_hparams` to have control over the log_dir
+                exp, ssi, sei = hparams(value.hparam_dict, metric_dict=value.metric_dict)
```
Comment (araffin) [inline stable_baselines3/common/logger.py:419]: please use explicit names, what is ssi? sei?

File: tests/test_logger.py [hunk 2]
```diff
@@ -296,6 +297,19 @@ def test_report_figure_to_unsupported_format_raises_error(tmp_path, unsupported_
     writer.close()
 
 
+@pytest.mark.parametrize("unsupported_format", ["stdout", "log", "json", "csv"])
```
Comment (araffin) [inline tests/test_logger.py:300]: is the new feature also tested somewhere?
currently only the failure case is tested?

File: stable_baselines3/common/logger.py [hunk 3]
```diff
@@ -66,6 +67,22 @@ def __init__(self, image: Union[th.Tensor, np.ndarray, str], dataformats: str):
...
+
+    :param hparam_dict: key-value pairs of hyperparameters to log
+    :param metric_dict: key-value pairs of metrics to log
+        A non-empty metrics dict is required to display hyperparameters in the corresponding Tensorboard section.
+    """
+
+    def __init__(self, hparam_dict: Dict[str, Union[bool, str, float, int, None]], metric_dict: Dict[str, Union[float, int]]):
+        self.hparam_dict = hparam_dict
+        if not metric_dict:
+            warnings.warn("`metric_dict` must not be empty to display hyperparameters to the HPARAMS tensorboard tab.")
```
Comment (araffin) [inline stable_baselines3/common/logger.py:82]: why not an error in that case?

File: stable_baselines3/common/logger.py (hunk 1 above)
Comment (timothe-chaumont) [reply stable_baselines3/common/logger.py]: It's done:
```python
experiment, session_start_info, session_end_info = hparams(value.hparam_dict, metric_dict=value.metric_dict)
```

The content and meaning of those variables is described in [the function's docstring](https://github.com/pytorch/pytorch/blob/cf5330977d10a3585358fb02c049939bf1401074/torch/utils/tensorboard/summary.py#L79-L82). (And the initial short names where those used in [pytorch SummaryWriter class](https://github.com/pytorch/pytorch/blob/master/torch/utils/tensorboard/writer.py#L336))

File: stable_baselines3/common/logger.py (hunk 3 above)
Comment (timothe-chaumont) [reply stable_baselines3/common/logger.py]: True, nothing will break but the hparams tab will be empty. I have added an exception.

File: tests/test_logger.py (hunk 2 above)
Comment (timothe-chaumont) [reply tests/test_logger.py:300]: I tested the new feature in practice, but not programmatically. It is not as easy as for the other types of logs (e.g. images) :
> I did not create a test_report_hparam_to_tensorboard test, as hparams logs are not seen by EventAccumulator (c.f. [Stack Overflow](https://stackoverflow.com/questions/70033253/read-out-tensorboard-hparams-programatically) and [EventAccumulator implementation](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/event_accumulator.py)).

Two solutions are proposed in [this stackoverflow page](https://stackoverflow.com/questions/70033253/read-out-tensorboard-hparams-programatically) but one requires an external library, and the other is not very clean.

Comment (araffin) [reply tests/test_logger.py:300]: more than reading the logged hparams, we should at least run the logger.

**Review by araffin — APPROVED** [review]
LGTM, thanks =)
